### PR TITLE
Dev 0.1.8

### PR DESCRIPTION
**Pull request by kamil-kaczmarek**
Changes:
* removed copy_transformer
* refactors in get_step_by_name
* corrected `_apply_suffix`
* Fix bug with names validation

Added:
* `DEFAULT_TRAINING_SETUP`
* new defaults is basically False everywhere
* `self._mode` and logic behind it
* set_mode_train, set_mode_inference, set_parameters_upstream_steps, reset functions

### CODE REVIEW

File: steppy/base.py [hunk 1]
```diff
@@ -376,8 +405,56 @@ def transform(self, data):
...
+            step_obj.persist_upstream_pipeline_structure = DEFAULT_TRAINING_SETUP['persist_upstream_pipeline_structure']
+        logger.info('Step {}, reset all upstream Steps to default training parameters, '
+                    'including this Step'.format(self.name))
+
+    def set_parameters_upstream_steps(self, parameters):
+        """Set parameters to all upstream Steps including this Step.
+        Parameters is dict() where key is Step attribute, and value is new value to set.
+        """
+        assert isinstance(parameters, dict), 'parameters must be dict, got {} instead'.format(type(parameters))
+        for name, step_obj in self.all_upstream_steps.items():
```
Comment (jakubczakon) [inline steppy/base.py]: @kamil-kaczmarek again name is not used anywhere in this loop. Probably one should use .values() .

File: steppy/base.py [hunk 2]
```diff
@@ -507,25 +596,19 @@ def _unpack(self, step_inputs):
...
-        copy_filepath = os.path.join(dirpath, 'transformers', name)
-        logger.info('copying transformer from {} to {}'.format(original_filepath, copy_filepath))
-        shutil.copyfile(original_filepath, copy_filepath)
-
     def _prepare_experiment_directories(self):
         if not os.path.exists(os.path.join(self.exp_dir, 'outputs')):
             logger.info('initializing experiment directories under {}'.format(self.exp_dir))
             for dir_name in ['transformers', 'outputs']:
                 os.makedirs(os.path.join(self.exp_dir, dir_name), exist_ok=True)
+            for dir_name in ['train', 'inference']:
```
Comment (jakubczakon) [inline steppy/base.py]: @kamil-kaczmarek user defined output option is needed.

File: steppy/base.py [hunk 3]
```diff
@@ -539,25 +622,28 @@ def _format_step_name(self, name, transformer):
             name = str(name)
         else:
             name = transformer.__class__.__name__
-        return '{}{}'.format(name, self._get_step_suffix(name))
+        return name
 
-    def _validate_step_name(self, name):
+    @staticmethod
+    def _validate_step_name(name):
```
Comment (jakubczakon) [inline steppy/base.py]: private staticmethod seems weird I would choose either one or the other.

File: steppy/base.py [hunk 4]
```diff
@@ -376,8 +405,56 @@ def transform(self, data):
...
+            'is_fittable': True,
+            'force_fitting': True,
+            'persist_output': False,
+            'cache_output': False,
+            'load_persisted_output': False,
+            'persist_upstream_pipeline_structure': False
+        """
+        self.clean_cache_upstream_steps()
+        self.set_mode_train()
+        for name, step_obj in self.all_upstream_steps.items():
```
[the remark this answers is not in the document]
Comment (kamil-kaczmarek) [reply steppy/base.py]: corrected

File: steppy/base.py [hunk 5]
```diff
@@ -376,8 +405,56 @@ def transform(self, data):
...
+            step_obj.is_fittable = DEFAULT_TRAINING_SETUP['is_fittable']
+            step_obj.force_fitting = DEFAULT_TRAINING_SETUP['force_fitting']
+            step_obj.persist_output = DEFAULT_TRAINING_SETUP['persist_output']
+            step_obj.cache_output = DEFAULT_TRAINING_SETUP['cache_output']
+            step_obj.load_persisted_output = DEFAULT_TRAINING_SETUP['load_persisted_output']
+            step_obj.persist_upstream_pipeline_structure = DEFAULT_TRAINING_SETUP['persist_upstream_pipeline_structure']
+        logger.info('Step {}, reset all upstream Steps to default training parameters, '
+                    'including this Step'.format(self.name))
+
+    def set_parameters_upstream_steps(self, parameters):
```
[the remark this answers is not in the document]
Comment (kamil-kaczmarek) [reply steppy/base.py]: agree, corrected

File: steppy/base.py (hunk 1 above)
Comment (kamil-kaczmarek) [reply steppy/base.py]: corrected

File: steppy/base.py (hunk 3 above)
Comment (kamil-kaczmarek) [reply steppy/base.py]: agree, use private method

File: steppy/base.py (hunk 2 above)
Comment (kamil-kaczmarek) [reply steppy/base.py]: Yes, I'll implement it tomorrow.

File: steppy/base.py [hunk 6]
```diff
@@ -569,6 +654,15 @@ def _build_structure_dict(self, structure_dict):
             structure_dict['edges'].add((input_data, self.name))
         return structure_dict
 
+    def _register_step(self):
```
Comment (jakubczakon) [inline steppy/base.py]: Since it is not used (yet) it should probably be deleted

File: steppy/base.py [hunk 7]
```diff
@@ -569,6 +654,15 @@ def _build_structure_dict(self, structure_dict):
...
+        pass
+
+    def _set_mode(self, mode):
```
Comment (jakubczakon) [inline steppy/base.py]: I could potentially _set_mode('inference') and run transform on the train set in which case all the cache could be useful and so now that I think about it, I am not sure about the clean_cache_upstream

File: steppy/base.py [hunk 8]
```diff
@@ -34,17 +43,17 @@ class Step:
...
             stored.
             Default is ``~/.steppy``.
-            The following sub-directories will be created, if they were not created by
-            other Steps:
+            The following directories will be created under ``~/.steppy``, if they were not created by
+            preceding Steps:
 
-            * transformers: transformer objects are persisted in this folder
-            * outputs:      step output dictionaries are persisted in this folder
+            * transformer: transformer objects are persisted in this folder
```
Comment (jakubczakon) [inline steppy/base.py]: typo trasnormerS

File: steppy/base.py [hunk 9]
```diff
@@ -204,14 +213,11 @@ def __init__(self,
...
-        assert isinstance(persist_upstream_pipeline_structure, bool),\
-            'Step {} error, persist_upstream_pipeline_structure must be bool, got {} instead.' \
-            .format(name, type(persist_upstream_pipeline_structure))
 
         logger.info('Initializing Step {}'.format(name))
 
-        self.name = name
         self.transformer = transformer
+        self.output_directory = output_directory
         self.input_steps = input_steps or []
```
Comment (jakubczakon) [inline steppy/base.py]: Cool, I like this

File: steppy/base.py (hunk 8 above)
Comment (kamil-kaczmarek) [reply steppy/base.py]: corrected

**Pull request merged**